analysis/parameter_tuning.py

```diff
@@ -190,22 +192,28 @@ def _find_candidates_constant_relative_step(histogram: histograms.Histogram,
                                             max_candidates: int) -> List[int]:
     """Implementation of CONSTANT_RELATIVE_STEP strategy."""
     max_value = histogram.max_value
-    # relative step varies from 1% to 0.1%
-    # because generate_possible_contribution_bounds generate bounds by changing
-    # only up to first 3 digits, for example 100000, 101000, 102000... Then
-    # relative step between neighbouring elements
-    # varies (101000 - 100000) / 100000 = 0.01 and
-    # (1000000 - 999000) / 999000 ~= 0.001.
-    candidates = private_contribution_bounds.generate_possible_contribution_bounds(
-        max_value)
-    n_max_without_max_value = max_candidates - 1
-    if len(candidates) > n_max_without_max_value:
-        delta = len(candidates) / n_max_without_max_value
-        candidates = [
-            candidates[int(i * delta)] for i in range(n_max_without_max_value)
-        ]
-    if candidates[-1] != max_value:
-        candidates.append(max_value)
+    if max_value < 1:
+        raise Exception("max_value has to be >= 1.")
+    max_candidates = min(max_candidates, max_value)
+    if max_candidates <= 0:
+        raise Exception("max_candidates have to be positive")
+    elif max_candidates == 1:
+        return [1]
+    step = pow(max_value, 1 / (max_candidates - 1))
+    candidates = [1]
+    accumulated = candidates[-1]
+    for i in range(1, max_candidates):
+        previous_candidate = candidates[-1]
+        if previous_candidate >= max_value:
```

Comment:
this `if` will probably never happen, but I can't prove so I would keep it :)

Reply:
It can happen `if max_value < max_candidates` and this is a valid case to generate smaller number of candidates

Reply:
no, it won't happen because in this case we will generate all values (step = 1) and for loop will end before this "if" turns to true.